### PR TITLE
Cleanups for JSDesugaring

### PR DESCRIPTION
**Pull request by gzm0**

### CODE REVIEW

File: project/BinaryIncompatibilities.scala [hunk 1]
```diff
@@ -23,7 +23,49 @@ object BinaryIncompatibilities {
...
-          "org.scalajs.core.tools.linker.backend.emitter.ScalaJSClassEmitter.genClassDef")
+          "org.scalajs.core.tools.linker.backend.emitter.ScalaJSClassEmitter.genClassDef"),
+      ProblemFilters.exclude[DirectMissingMethodProblem](
+          "org.scalajs.core.tools.linker.backend.emitter.JSDesugaring.genCallHelper"),
+      ProblemFilters.exclude[DirectMissingMethodProblem](
+          "org.scalajs.core.tools.linker.backend.emitter.JSDesugaring.genLet"),
+      ProblemFilters.exclude[DirectMissingMethodProblem](
+          "org.scalajs.core.tools.linker.backend.emitter.JSDesugaring.genRawJSClassConstructor"),
+      ProblemFilters.exclude[DirectMissingMethodProblem](
+          "org.scalajs.core.tools.linker.backend.emitter.JSDesugaring.genRawJSClassConstructor"),
```
Comment (sjrd) [inline project/BinaryIncompatibilities.scala]: There are duplicates, here and below.

File: project/BinaryIncompatibilities.scala [hunk 2]
```diff
@@ -23,7 +23,49 @@ object BinaryIncompatibilities {
...
+      ProblemFilters.exclude[DirectMissingMethodProblem](
+          "org.scalajs.core.tools.linker.backend.emitter.JSDesugaring.envFieldDef"),
+      ProblemFilters.exclude[DirectMissingMethodProblem](
+          "org.scalajs.core.tools.linker.backend.emitter.JSDesugaring.genIsInstanceOf"),
+      ProblemFilters.exclude[DirectMissingMethodProblem](
+          "org.scalajs.core.tools.linker.backend.emitter.JSDesugaring.this"),
+      ProblemFilters.exclude[DirectMissingMethodProblem](
+          "org.scalajs.core.tools.linker.backend.emitter.ScalaJSClassEmitter.outputMode"),
+      ProblemFilters.exclude[DirectMissingMethodProblem](
+          "org.scalajs.core.tools.linker.backend.emitter.ScalaJSClassEmitter.semantics"),
```
Comment (sjrd) [inline project/BinaryIncompatibilities.scala]: Could you keep these two about `ScalaJSClassEmitter.*` together with the other ones at the top of this list?

File: tools/shared/src/main/scala/org/scalajs/core/tools/linker/backend/emitter/JSDesugaring.scala [hunk 3]
```diff
@@ -218,20 +218,24 @@ private[emitter] class JSDesugaring(semantics: Semantics,
...
   }
 
   /** Desugars a statement or an expression. */
   private[emitter] def desugarTree(
-      enclosingClassName: String,
+      enclosingClassName: Option[String],
       tree: Tree, isStat: Boolean)(
-      implicit globalKnowledge: GlobalKnowledge): js.Tree = {
-    val desugar = new JSDesugar(enclosingClassName)
+    implicit globalKnowledge: GlobalKnowledge): js.Tree = {
```
Comment (sjrd) [inline tools/shared/src/main/scala/org/scalajs/core/tools/linker/backend/emitter/JSDesugaring.scala]: Indentation issue.

File: tools/shared/src/main/scala/org/scalajs/core/tools/linker/backend/emitter/JSDesugaring.scala [hunk 4]
```diff
@@ -240,8 +244,7 @@ private[emitter] class JSDesugaring(semantics: Semantics,
   private[emitter] def transformParamDef(paramDef: ParamDef): js.ParamDef =
     js.ParamDef(paramDef.name, paramDef.rest)(paramDef.pos)
 
-  private class JSDesugar(enclosingClassName: String)(
-      implicit globalKnowledge: GlobalKnowledge) {
+  private class JSDesugar(implicit globalKnowledge: GlobalKnowledge) {
```
Comment (sjrd) [inline tools/shared/src/main/scala/org/scalajs/core/tools/linker/backend/emitter/JSDesugaring.scala]: Keep an explicit empty pair of `()` in the constructor. Scala adds them implicitly for a constructor that has only an implicit parameter list, but IMO they should be explicit.

**Comment by gzm0** [comment]
Updated

**Comment by sjrd** [comment]
It got new MiMa errors. Maybe needs a rebase after #2716 was merged?

**Comment by gzm0** [comment]
Updated.

**Pull request merged**